src/acl/Options.h

```diff
@@ -13,51 +13,84 @@
 #include "sbuf/forward.h"
 
 #include <iosfwd>
-#include <map>
-#include <set>
+#include <vector>
 
-// After all same-name acl configuration lines are merged into one ACL:
-//   configuration = acl name type [option...] [[flag...] parameter...]
-//   option = -x[=value] | --name[=value]
-//   flag = option
+// After line continuation is handled by the preprocessor, an ACL object
+// configuration can be visualized as a sequence of same-name "acl ..." lines:
 //
-// Options and flags use the same syntax, but differ in scope and handling code:
-// * ACL options appear before all parameters and apply to all parameters.
-//   They are handled by ACL kids (or equivalent).
-// * Parameter flags may appear after some other parameters and apply only to
-//   the subsequent parameters (until they are overwritten by later flags).
-//   They are handled by ACLData kids.
-// ACL options parsing code skips and leaves leading parameter flags (if any)
-// for ACLData code to process.
+// L1: acl exampleA typeT parameter1 -i parameter2 parameter3
+// L2: acl exampleA typeT parameter4
+// L3: acl exampleA typeT -i -n parameter5 +i parameter6
+// L4: acl exampleA typeT -n parameter7
+//
+// There are two kinds of ACL options (a.k.a. flags):
+//
+// * Global (e.g., "-n"): Applies to all parameters regardless of where the
+//   option was discovered/parsed (e.g., "-n" on L3 affects parameter2 on L1).
+//   Declared by ACL class kids (or equivalent) via ACL::options().
+//
+// * Line (e.g., "-i"): Applies to the yet unparsed ACL parameters of the
+//   current "acl ..." line (e.g., "-i" on L1 has no effect on parameter4 on L2)
+//   Declared by ACLData class kids (or equivalent) via lineOptions().
+//
+// Here is the option:explicitly-affected-parameters map for the above exampleA:
+//   "-n": parameter1-7 (i.e. all parameters)
+//   "-i": parameter2, parameter3; parameter5
+//   "+i": parameter6
+//
+// The option name spelling determines the option kind and effect.
+// Both option kinds use the same general option configuration syntax:
+//   option = name[=value]
+// where "name" is option-specific spelling that looks like -x, +x, or --long
+//
+// On each "acl ..." line, global options can only appear before the first
+// parameter, while line options can go before any parameter.
```

Comment:
IMO we should look at deprecating the ability to put options in the middle of ACL values. Admin can trivially split ACL definitions into different lines if they need the line-specific options for different groups of values.

We have here an opportunity with this redesign to document that options should always go after the type field and have the parser log an "UPGRADE WARNING" when it finds old config using -i/+i etc after the first value of a line.

This will make it easier in future to implement line association for the config dumper.

Reply:
> Admin can trivially split ACL definitions into different lines

Such splitting will not help when loading ACL parameters from parameter files, which is a very common use case. Such files often contain line options. When parameters are loaded from ACL parameter files, multiple file lines are treated as a single `acl... ` line. Changing that approach is very difficult (I can explain why if needed). Within the existing approach, the low-level parameter file parser does not have access to the higher-level ACL parser which knows about the line boundaries and can call unconfigure() between the lines inside parameter files. That low-level parser is also used for non-ACL parameters...

Let's not bring this complex change into this PR. It can be added later IF we reach consensus that it is a good idea and find a way to do it well. This PR does not make that future change more difficult. In fact, it may make it a bit easier.

Reply:
That file loading is something we are going to have to discuss and sort out definitions for. I am fine with leaving it out of this PR though.